File: packages/python/src/fraud_catcher/algorithms/time.py
"""
Time-based fraud detection algorithm.
"""

import math
from datetime import datetime
from typing import Dict, List, Optional, Any
from dataclasses import dataclass
from ..models import Transaction, DetectionRule, Location


@dataclass
class TimeConfig:
    """Configuration for time algorithm."""
    suspicious_hours: List[int]  # Hours considered suspicious (0-23)
    weekend_risk_multiplier: float
    holiday_risk_multiplier: float
    timezone_threshold: int  # Hours difference to consider suspicious
    enable_holiday_detection: bool
    custom_holidays: Optional[List[datetime]] = None


@dataclass
class TimePattern:
    """Time pattern information."""
    hour: int
    day_of_week: int  # 0 = Sunday, 1 = Monday, etc.
    is_weekend: bool
    is_holiday: bool
    timezone: str
    risk_score: float


class TimeAlgorithm:
    """Detects fraud based on transaction timing patterns."""
    
    def __init__(self, config: TimeConfig):
        self.config = config
        self.user_time_patterns: Dict[str, List[TimePattern]] = {}
        self.holidays: set = set()
        self._initialize_holidays()
    
    async def analyze(self, transaction: Transaction, rule: DetectionRule) -> float:
        """Analyze transaction for time-based fraud patterns."""
        transaction_time = transaction.timestamp if isinstance(transaction.timestamp, datetime) else datetime.now()
        time_pattern = self._analyze_time_pattern(transaction_time, transaction)
        
        risk_score = 0.0
        
        # Check for suspicious hours
        if time_pattern.hour in self.config.suspicious_hours:
            risk_score += 0.4
        
        # Check for weekend transactions
        if time_pattern.is_weekend:
            risk_score += 0.2 * self.config.weekend_risk_multiplier
        
        # Check for holiday transactions
        if time_pattern.is_holiday:
            risk_score += 0.3 * self.config.holiday_risk_multiplier
        
        # Check for unusual time patterns for this user
        user_pattern_risk = self._analyze_user_time_pattern(transaction.user_id, time_pattern)
        risk_score += user_pattern_risk
        
        # Check for timezone anomalies
        timezone_risk = self._analyze_timezone_anomaly(transaction, time_pattern)
        risk_score += timezone_risk
        
        # Store pattern for future analysis
        self._store_user_time_pattern(transaction.user_id, time_pattern)
        
        return min(risk_score, 1.0)
    
    def _analyze_time_pattern(self, transaction_time: datetime, transaction: Transaction) -> TimePattern:
        """Analyze time pattern from transaction."""
        hour = transaction_time.hour
        day_of_week = transaction_time.weekday() + 1  # Convert to 1-7 (Monday=1, Sunday=7)
        is_weekend = day_of_week in [6, 7]  # Saturday or Sunday
        is_holiday = self._is_holiday(transaction_time)
        timezone = self._extract_timezone(transaction)
        
        return TimePattern(
            hour=hour,
            day_of_week=day_of_week,
            is_weekend=is_weekend,
            is_holiday=is_holiday,
            timezone=timezone,
            risk_score=0.0
        )
    
    def _analyze_user_time_pattern(self, user_id: str, current_pattern: TimePattern) -> float:
        """Analyze user's time patterns for anomalies."""
        user_patterns = self.user_time_patterns.get(user_id, [])
        
        if not user_patterns:
            return 0.1  # Slight risk for first transaction
        
        # Analyze frequency of transactions at this time
        similar_patterns = [
            pattern for pattern in user_patterns
            if pattern.hour == current_pattern.hour and pattern.day_of_week == current_pattern.day_of_week
        ]
        
        total_patterns = len(user_patterns)
        similar_count = len(similar_patterns)
        frequency = similar_count / total_patterns if total_patterns > 0 else 0
        
        # If user rarely transacts at this time, it's suspicious
        if frequency < 0.1:
            return 0.3
        elif frequency < 0.3:
            return 0.1
        
        return 0.0
    
    def _analyze_timezone_anomaly(self, transaction: Transaction, time_pattern: TimePattern) -> float:
        """Analyze timezone anomalies."""
        if not transaction.location:
            return 0.0  # No location data
        
        # Extract timezone from location or transaction metadata
        expected_timezone = self._get_timezone_from_location(transaction.location)
        actual_timezone = time_pattern.timezone
        
        if expected_timezone and actual_timezone:
            timezone_diff = self._calculate_timezone_difference(expected_timezone, actual_timezone)
            if timezone_diff > self.config.timezone_threshold:
                return 0.4  # High risk for timezone mismatch
        
        return 0.0
    
    def _extract_timezone(self, transaction: Transaction) -> str:
        """Extract timezone from transaction."""
        if transaction.metadata and 'timezone' in transaction.metadata:
            return transaction.metadata['timezone']
        
        if transaction.location and transaction.location.country:
            return self._get_country_timezone(transaction.location.country)
        
        return 'UTC'
    
    def _get_timezone_from_location(self, location: Location) -> Optional[str]:
        """Get timezone from location."""
        if location.country:
            return self._get_country_timezone(location.country)
        return None
    
    def _get_country_timezone(self, country: str) -> str:
        """Get timezone for country."""
        timezone_map = {
            'US': 'America/New_York',
            'GB': 'Europe/London',
            'DE': 'Europe/Berlin',
            'FR': 'Europe/Paris',
            'JP': 'Asia/Tokyo',
            'AU': 'Australia/Sydney',
            'CA': 'America/Toronto',
            'BR': 'America/Sao_Paulo',
            'IN': 'Asia/Kolkata',
            'CN': 'Asia/Shanghai'
        }
        return timezone_map.get(country, 'UTC')
    
    def _calculate_timezone_difference(self, tz1: str, tz2: str) -> int:
        """Calculate timezone difference in hours."""
        timezone_offsets = {
            'UTC': 0,
            'America/New_York': -5,
            'Europe/London': 0,
            'Europe/Berlin': 1,
            'Europe/Paris': 1,
            'Asia/Tokyo': 9,
            'Australia/Sydney': 10,
            'America/Toronto': -5,
            'America/Sao_Paulo': -3,
            'Asia/Kolkata': 5.5,
            'Asia/Shanghai': 8
        }
        
        offset1 = timezone_offsets.get(tz1, 0)
        offset2 = timezone_offsets.get(tz2, 0)
        
        return abs(int(offset1 - offset2))
    
    def _is_holiday(self, date: datetime) -> bool:
        """Check if date is a holiday."""
        if not self.config.enable_holiday_detection:
            return False
        
        date_str = date.strftime('%Y-%m-%d')
        
        # Check custom holidays
        if self.config.custom_holidays:
            for holiday in self.config.custom_holidays:
                if holiday.strftime('%Y-%m-%d') == date_str:
                    return True
        
        # Check built-in holidays
        return date_str in self.holidays
    
    def _initialize_holidays(self) -> None:
        """Initialize holiday list."""
        if not self.config.enable_holiday_detection:
            return
        
        current_year = datetime.now().year
        holidays = [
            f'{current_year}-01-01',  # New Year's Day
            f'{current_year}-12-25',  # Christmas
            f'{current_year}-12-31',  # New Year's Eve
        ]
        
        self.holidays = set(holidays)
    
    def _store_user_time_pattern(self, user_id: str, pattern: TimePattern) -> None:
        """Store time pattern for user."""
        if user_id not in self.user_time_patterns:
            self.user_time_patterns[user_id] = []
        
        self.user_time_patterns[user_id].append(pattern)
        
        # Keep only recent patterns to manage memory
        if len(self.user_time_patterns[user_id]) > 100:
            self.user_time_patterns[user_id] = self.user_time_patterns[user_id][-100:]
    
    # Utility methods
    def get_user_time_patterns(self, user_id: str) -> List[TimePattern]:
        """Get time patterns for user."""
        return self.user_time_patterns.get(user_id, [])
    
    def get_most_common_transaction_time(self, user_id: str) -> Optional[Dict[str, int]]:
        """Get most common transaction time for user."""
        patterns = self.user_time_patterns.get(user_id, [])
        if not patterns:
            return None
        
        time_counts = {}
        for pattern in patterns:
            key = f"{pattern.hour}-{pattern.day_of_week}"
            time_counts[key] = time_counts.get(key, 0) + 1
        
        if not time_counts:
            return None
        
        most_common = max(time_counts.items(), key=lambda x: x[1])
        hour, day_of_week = most_common[0].split('-')
        
        return {
            'hour': int(hour),
            'day_of_week': int(day_of_week)
        }
    
    def is_suspicious_time(self, hour: int, day_of_week: int) -> bool:
        """Check if time is suspicious."""
        return (hour in self.config.suspicious_hours or 
                day_of_week in [6, 7])  # Weekend
    
    def get_time_risk_level(self, hour: int, day_of_week: int) -> str:
        """Get risk level for time."""
        if hour in self.config.suspicious_hours:
            return 'high'
        elif day_of_week in [6, 7]:  # Weekend
            return 'medium'
        else:
            return 'low'

File: packages/go/algorithms/time.go
package algorithms

import (
	"context"
	"fmt"
	"math"
	"sync"
	"time"

	fraudcatcher "github.com/fraud-catcher/go"
)

// TimeConfig represents the configuration for the time algorithm
type TimeConfig struct {
	SuspiciousHours        []int       `json:"suspicious_hours"`
	WeekendRiskMultiplier  float64     `json:"weekend_risk_multiplier"`
	HolidayRiskMultiplier  float64     `json:"holiday_risk_multiplier"`
	TimezoneThreshold      int         `json:"timezone_threshold"`
	EnableHolidayDetection bool        `json:"enable_holiday_detection"`
	CustomHolidays         []time.Time `json:"custom_holidays"`
	EnableUserTimeProfiles bool        `json:"enable_user_time_profiles"`
}

// TimePattern represents a time pattern for analysis
type TimePattern struct {
	Hour      int     `json:"hour"`
	DayOfWeek int     `json:"day_of_week"`
	IsWeekend bool    `json:"is_weekend"`
	IsHoliday bool    `json:"is_holiday"`
	Timezone  string  `json:"timezone"`
	RiskScore float64 `json:"risk_score"`
}

// UserTimeProfile represents a user's time-based transaction profile
type UserTimeProfile struct {
	UserID           string        `json:"user_id"`
	TransactionCount int           `json:"transaction_count"`
	HourDistribution map[int]int   `json:"hour_distribution"`
	DayDistribution  map[int]int   `json:"day_distribution"`
	WeekendActivity  int           `json:"weekend_activity"`
	HolidayActivity  int           `json:"holiday_activity"`
	PreferredHours   []int         `json:"preferred_hours"`
	LastTransaction  time.Time     `json:"last_transaction"`
	TimePatterns     []TimePattern `json:"time_patterns"`
	LastUpdated      int64         `json:"last_updated"`
}

// TimeAlgorithm implements fraud detection based on transaction timing
type TimeAlgorithm struct {
	*fraudcatcher.BaseAlgorithm
	config       TimeConfig
	userProfiles map[string]*UserTimeProfile
	holidays     map[string]bool
	mutex        sync.RWMutex
}

// NewTimeAlgorithm creates a new time algorithm
func NewTimeAlgorithm(config TimeConfig) *TimeAlgorithm {
	base := fraudcatcher.NewBaseAlgorithm("time", "Detects suspicious transaction timing patterns", 0.6)

	algorithm := &TimeAlgorithm{
		BaseAlgorithm: base,
		config:        config,
		userProfiles:  make(map[string]*UserTimeProfile),
		holidays:      make(map[string]bool),
	}

	algorithm.initializeHolidays()
	return algorithm
}

// Analyze analyzes a transaction for time-based fraud
func (ta *TimeAlgorithm) Analyze(ctx context.Context, transaction fraudcatcher.Transaction, rule fraudcatcher.DetectionRule) (float64, error) {
	if !ta.IsEnabled() {
		return 0.0, nil
	}

	transactionTime := transaction.Timestamp
	timePattern := ta.analyzeTimePattern(transactionTime, transaction)

	riskScore := 0.0

	// Check for suspicious hours
	if ta.isSuspiciousHour(timePattern.Hour) {
		riskScore += 0.4
	}

	// Check for weekend transactions
	if timePattern.IsWeekend {
		riskScore += 0.2 * ta.config.WeekendRiskMultiplier
	}

	// Check for holiday transactions
	if timePattern.IsHoliday {
		riskScore += 0.3 * ta.config.HolidayRiskMultiplier
	}

	// Check for unusual time patterns for this user
	if ta.config.EnableUserTimeProfiles {
		userPatternRisk := ta.analyzeUserTimePattern(transaction.UserID, timePattern)
		riskScore += userPatternRisk
	}

	// Check for timezone anomalies
	timezoneRisk := ta.analyzeTimezoneAnomaly(transaction, timePattern)
	riskScore += timezoneRisk

	// Store pattern for future analysis
	ta.storeUserTimePattern(transaction.UserID, timePattern)

	return math.Min(riskScore, 1.0), nil
}

// analyzeTimePattern analyzes the time pattern of a transaction
func (ta *TimeAlgorithm) analyzeTimePattern(transactionTime time.Time, transaction fraudcatcher.Transaction) TimePattern {
	hour := transactionTime.Hour()
	dayOfWeek := int(transactionTime.Weekday())
	isWeekend := dayOfWeek == 0 || dayOfWeek == 6 // Sunday or Saturday
	isHoliday := ta.isHoliday(transactionTime)
	timezone := ta.extractTimezone(transaction)

	return TimePattern{
		Hour:      hour,
		DayOfWeek: dayOfWeek,
		IsWeekend: isWeekend,
		IsHoliday: isHoliday,
		Timezone:  timezone,
		RiskScore: 0.0,
	}
}

// isSuspiciousHour checks if an hour is considered suspicious
func (ta *TimeAlgorithm) isSuspiciousHour(hour int) bool {
	for _, suspiciousHour := range ta.config.SuspiciousHours {
		if hour == suspiciousHour {
			return true
		}
	}
	return false
}

// analyzeUserTimePattern analyzes user-specific time patterns
func (ta *TimeAlgorithm) analyzeUserTimePattern(userID string, currentPattern TimePattern) float64 {
	profile := ta.getUserProfile(userID)

	if profile.TransactionCount == 0 {
		return 0.1 // Slight risk for first transaction
	}

	// Analyze frequency of transactions at this time
	similarPatterns := ta.findSimilarPatterns(profile, currentPattern)

	totalPatterns := len(profile.TimePatterns)
	similarCount := len(similarPatterns)
	frequency := float64(similarCount) / float64(totalPatterns)

	// If user rarely transacts at this time, it's suspicious
	if frequency < 0.1 {
		return 0.3
	} else if frequency < 0.3 {
		return 0.1
	}

	return 0.0
}

// findSimilarPatterns finds patterns similar to the current pattern
func (ta *TimeAlgorithm) findSimilarPatterns(profile *UserTimeProfile, currentPattern TimePattern) []TimePattern {
	var similar []TimePattern

	for _, pattern := range profile.TimePatterns {
		if pattern.Hour == currentPattern.Hour && pattern.DayOfWeek == currentPattern.DayOfWeek {
			similar = append(similar, pattern)
		}
	}

	return similar
}

// analyzeTimezoneAnomaly analyzes timezone-related anomalies
func (ta *TimeAlgorithm) analyzeTimezoneAnomaly(transaction fraudcatcher.Transaction, timePattern TimePattern) float64 {
	if transaction.Location == nil {
		return 0.0 // No location data
	}

	// Extract timezone from location or transaction metadata
	expectedTimezone := ta.getTimezoneFromLocation(transaction.Location)
	actualTimezone := timePattern.Timezone

	if expectedTimezone != "" && actualTimezone != "" {
		timezoneDiff := ta.calculateTimezoneDifference(expectedTimezone, actualTimezone)
		if timezoneDiff > ta.config.TimezoneThreshold {
			return 0.4 // High risk for timezone mismatch
		}
	}

	return 0.0
}

// extractTimezone extracts timezone from transaction
func (ta *TimeAlgorithm) extractTimezone(transaction fraudcatcher.Transaction) string {
	// Try to extract timezone from metadata
	if transaction.Metadata != nil {
		if tz, exists := transaction.Metadata["timezone"]; exists {
			if tzStr, ok := tz.(string); ok {
				return tzStr
			}
		}
	}

	// Try to extract from location
	if transaction.Location != nil && transaction.Location.Country != "" {
		return ta.getCountryTimezone(transaction.Location.Country)
	}

	// Default to UTC
	return "UTC"
}

// getTimezoneFromLocation gets timezone from location
func (ta *TimeAlgorithm) getTimezoneFromLocation(location *fraudcatcher.Location) string {
	if location.Country != "" {
		return ta.getCountryTimezone(location.Country)
	}
	return ""
}

// getCountryTimezone returns timezone for a country
func (ta *TimeAlgorithm) getCountryTimezone(country string) string {
	timezoneMap := map[string]string{
		"US": "America/New_York",
		"GB": "Europe/London",
		"DE": "Europe/Berlin",
		"FR": "Europe/Paris",
		"JP": "Asia/Tokyo",
		"AU": "Australia/Sydney",
		"CA": "America/Toronto",
		"BR": "America/Sao_Paulo",
		"IN": "Asia/Kolkata",
		"CN": "Asia/Shanghai",
	}

	return timezoneMap[country]
}

// calculateTimezoneDifference calculates timezone difference in hours
func (ta *TimeAlgorithm) calculateTimezoneDifference(tz1, tz2 string) int {
	timezoneOffsets := map[string]int{
		"UTC":               0,
		"America/New_York":  -5,
		"Europe/London":     0,
		"Europe/Berlin":     1,
		"Europe/Paris":      1,
		"Asia/Tokyo":        9,
		"Australia/Sydney":  10,
		"America/Toronto":   -5,
		"America/Sao_Paulo": -3,
		"Asia/Kolkata":      5,
		"Asia/Shanghai":     8,
	}

	offset1 := timezoneOffsets[tz1]
	offset2 := timezoneOffsets[tz2]

	diff := offset1 - offset2
	if diff < 0 {
		diff = -diff
	}

	return diff
}

// isHoliday checks if a date is a holiday
func (ta *TimeAlgorithm) isHoliday(date time.Time) bool {
	if !ta.config.EnableHolidayDetection {
		return false
	}

	dateStr := date.Format("2006-01-02")

	// Check custom holidays
	for _, holiday := range ta.config.CustomHolidays {
		if holiday.Format("2006-01-02") == dateStr {
			return true
		}
	}

	// Check built-in holidays
	return ta.holidays[dateStr]
}

// initializeHolidays initializes built-in holidays
func (ta *TimeAlgorithm) initializeHolidays() {
	if !ta.config.EnableHolidayDetection {
		return
	}

	currentYear := time.Now().Year()
	holidays := []string{
		"2006-01-01", // New Year's Day
		"2006-12-25", // Christmas
		"2006-12-31", // New Year's Eve
	}

	for _, holiday := range holidays {
		// Replace year with current year
		holidayWithYear := holiday
		if holiday[0:4] == "2006" {
			holidayWithYear = time.Now().Format("2006") + holiday[4:]
		}
		ta.holidays[holidayWithYear] = true
	}
}

// getUserProfile returns the user profile for the given user ID
func (ta *TimeAlgorithm) getUserProfile(userID string) *UserTimeProfile {
	ta.mutex.Lock()
	defer ta.mutex.Unlock()

	profile, exists := ta.userProfiles[userID]
	if !exists {
		profile = &UserTimeProfile{
			UserID:           userID,
			HourDistribution: make(map[int]int),
			DayDistribution:  make(map[int]int),
			TimePatterns:     []TimePattern{},
			LastUpdated:      0,
		}
		ta.userProfiles[userID] = profile
	}

	return profile
}

// storeUserTimePattern stores a time pattern for a user
func (ta *TimeAlgorithm) storeUserTimePattern(userID string, pattern TimePattern) {
	profile := ta.getUserProfile(userID)

	ta.mutex.Lock()
	defer ta.mutex.Unlock()

	profile.TransactionCount++
	profile.HourDistribution[pattern.Hour]++
	profile.DayDistribution[pattern.DayOfWeek]++

	if pattern.IsWeekend {
		profile.WeekendActivity++
	}

	if pattern.IsHoliday {
		profile.HolidayActivity++
	}

	profile.TimePatterns = append(profile.TimePatterns, pattern)
	profile.LastTransaction = time.Now()
	profile.LastUpdated = time.Now().Unix()

	// Keep only last 100 patterns per user to manage memory
	if len(profile.TimePatterns) > 100 {
		profile.TimePatterns = profile.TimePatterns[len(profile.TimePatterns)-100:]
	}
}

// IsSuspiciousTime checks if a time is suspicious
func (ta *TimeAlgorithm) IsSuspiciousTime(hour int, dayOfWeek int) bool {
	return ta.isSuspiciousHour(hour) || (dayOfWeek == 0 || dayOfWeek == 6)
}

// GetTimeRiskLevel returns the risk level for a time
func (ta *TimeAlgorithm) GetTimeRiskLevel(hour int, dayOfWeek int) fraudcatcher.RiskLevel {
	if ta.isSuspiciousHour(hour) {
		return fraudcatcher.RiskHigh
	}

	if dayOfWeek == 0 || dayOfWeek == 6 {
		return fraudcatcher.RiskMedium
	}

	return fraudcatcher.RiskLow
}

// GetUserTimeProfile returns the user time profile
func (ta *TimeAlgorithm) GetUserTimeProfile(userID string) *UserTimeProfile {
	ta.mutex.RLock()
	defer ta.mutex.RUnlock()
	return ta.userProfiles[userID]
}

// GetMostCommonTransactionTime returns the most common transaction time for a user
func (ta *TimeAlgorithm) GetMostCommonTransactionTime(userID string) (int, int) {
	profile := ta.GetUserTimeProfile(userID)
	if profile == nil || len(profile.TimePatterns) == 0 {
		return -1, -1
	}

	timeCounts := make(map[string]int)

	for _, pattern := range profile.TimePatterns {
		key := fmt.Sprintf("%d-%d", pattern.Hour, pattern.DayOfWeek)
		timeCounts[key]++
	}

	maxCount := 0
	var mostCommon string

	for key, count := range timeCounts {
		if count > maxCount {
			maxCount = count
			mostCommon = key
		}
	}

	if mostCommon == "" {
		return -1, -1
	}

	// Parse the key back to hour and day
	var hour, day int
	fmt.Sscanf(mostCommon, "%d-%d", &hour, &day)
	return hour, day
}

// GetStats returns statistics for the time algorithm
func (ta *TimeAlgorithm) GetStats() map[string]interface{} {
	ta.mutex.RLock()
	defer ta.mutex.RUnlock()

	stats := ta.BaseAlgorithm.GetStats()
	stats["user_profiles_count"] = len(ta.userProfiles)
	stats["holidays_count"] = len(ta.holidays)
	stats["config"] = ta.config

	return stats
}

// Reset resets the algorithm state
func (ta *TimeAlgorithm) Reset() error {
	ta.mutex.Lock()
	defer ta.mutex.Unlock()

	ta.userProfiles = make(map[string]*UserTimeProfile)
	return nil
}
